geode-core/src/main/java/org/apache/geode/internal/net/BufferPool.java

```diff
@@ -90,46 +87,46 @@ static boolean computeUseDirectBuffers() {
    *
    * @return a byte buffer to be used for sending on this connection.
    */
-  public ByteBuffer acquireDirectSenderBuffer(int size) {
+  public PooledByteBuffer acquireDirectSenderBuffer(int size) {
     return acquireDirectBuffer(size, true);
   }
 
-  public ByteBuffer acquireDirectReceiveBuffer(int size) {
+  public PooledByteBuffer acquireDirectReceiveBuffer(int size) {
     return acquireDirectBuffer(size, false);
   }
 
   /**
    * try to acquire direct buffer, if enabled by configuration
    */
-  private ByteBuffer acquireDirectBuffer(int size, boolean send) {
-    ByteBuffer result;
+  private PooledByteBuffer acquireDirectBuffer(int size, boolean send) {
+    ByteBuffer byteBuffer;
 
     if (useDirectBuffers) {
       if (size <= MEDIUM_BUFFER_SIZE) {
-        result = acquirePredefinedFixedBuffer(send, size);
+        byteBuffer = acquirePredefinedFixedBuffer(send, size);
       } else {
-        result = acquireLargeBuffer(send, size);
+        byteBuffer = acquireLargeBuffer(send, size);
       }
-      if (result.capacity() > size) {
-        result.position(0).limit(size);
-        result = result.slice();
+      if (byteBuffer.capacity() > size) {
+        byteBuffer.position(0).limit(size);
+        return new PooledByteBuffer(byteBuffer, byteBuffer.slice());
```

Comment:
Going from 1 to 2 transient allocations here.

Reply:
I think these transient allocations are a drop in the bucket compared to the total number of transient allocations we have for each cache operation. And these transient instances are very small. I don't think it was worth using an internal JDK feature which is subject to change in any jdk release to avoid this transient.

Reply:
Also keep in mind that half of the users of BufferPool get a buffer and hold onto for an extended time. In those cases the PooledByteBuffer is not transient.

Reply:
The other half that don't invoke this method rapidly to get send buffers. While the size is small the effect of number of objects can be detrimental and if everyone along the path says it's small in the larger scope we eventually end up with noticeable latency. Great care has go into removing allocations like this in hot paths like this because they had measurable impact. While I agree p2p has many worse places for transient allocations I don't see it as proof to support a new one. I think we can avoid this completely by just returning the original `ByteBuffer`. To do so we need to stop slicing, which is not necessary.